xarray/backends/netCDF4_.py

```diff
@@ -25,14 +26,29 @@
                   '|': 'native'}
 
 
+class VariableWrapper(object):
+    """ Ensures that variable attributes and data access operates on open datastore """
+    def __init__(self, datastore, variable_name):
+        self.var = variable_name
+        self.datastore = datastore
+
+    def __getattr__(self, attr):
+        with self.datastore.ensure_open():
+            return getattr(self.datastore.ds.variables[self.var], attr)
+
+    def __getitem__(self, key):
+        with self.datastore.ensure_open():
+            return operator.getitem(self.datastore.ds.variables[self.var], key)
+
+
 class BaseNetCDF4Array(NDArrayMixin):
     def __init__(self, variable_name, datastore):
         self.datastore = datastore
         self.variable_name = variable_name
 
     @property
     def array(self):
-        return self.datastore.ds.variables[self.variable_name]
+        return VariableWrapper(self.datastore, self.variable_name)
```

Comment:
This will result in a lot of churn opening/closing files just to get things like the dtype for variables.

Instead, I would save `dtype` and `shape` when this object is created, and only use `ensure_open` when doing indexing. In other words, inherit from `NdimSizeLenMixin` instead of `NDArrayMixin` -- there's no particular reason why we need to keep the `array` property.